meta/state.go

```diff
@@ -0,0 +1,339 @@
+	r.raftLayer = newRaftLayer(s.RaftListener, s.Addr)
+
+	// Create a transport layer
+	r.transport = raft.NewNetworkTransport(r.raftLayer, 3, 10*time.Second, os.Stderr)
+
+	// Create peer storage.
+	r.peerStore = raft.NewJSONPeers(s.path, r.transport)
+
+	// Create the log store and stable store.
+	store, err := raftboltdb.NewBoltStore(filepath.Join(s.path, "raft.db"))
+	if err != nil {
+		return fmt.Errorf("new bolt store: %s", err)
+	}
+	r.raftStore = store
+
+	// Create the snapshot store.
+	snapshots, err := raft.NewFileSnapshotStore(s.path, raftSnapshotsRetained, os.Stderr)
+	if err != nil {
+		return fmt.Errorf("file snapshot store: %s", err)
+	}
+
+	// Create raft log.
+	ra, err := raft.NewRaft(config, (*storeFSM)(s), store, store, snapshots, r.peerStore, r.transport)
+	if err != nil {
+		return fmt.Errorf("new raft: %s", err)
+	}
+	r.raft = ra
+
+	return nil
+}
+
+func (r *localRaft) close() error {
+	// Shutdown raft.
+	if r.raft != nil {
+		r.raft.Shutdown()
+		r.raft = nil
+	}
+	if r.transport != nil {
+		r.transport.Close()
+		r.transport = nil
+	}
+	if r.raftStore != nil {
+		r.raftStore.Close()
+		r.raftStore = nil
+	}
+
+	return nil
+}
+
+func (r *localRaft) initialize() error {
+	s := r.store
+	// If we have committed entries then the store is already in the cluster.
+	if index, err := r.raftStore.LastIndex(); err != nil {
+		return fmt.Errorf("last index: %s", err)
+	} else if index > 0 {
+		return nil
+	}
+
+	// Force set peers.
+	if err := r.setPeers(s.peers); err != nil {
+		return fmt.Errorf("set raft peers: %s", err)
+	}
+
+	return nil
+}
+
+// apply applies a serialized command to the raft log.
+func (r *localRaft) apply(b []byte) error {
+	// Apply to raft log.
+	f := r.raft.Apply(b, 0)
+	if err := f.Error(); err != nil {
+		return err
+	}
+
+	// Return response if it's an error.
+	// No other non-nil objects should be returned.
+	resp := f.Response()
+	if err, ok := resp.(error); ok {
+		return lookupError(err)
+	}
+	assert(resp == nil, "unexpected response: %#v", resp)
+
+	return nil
+}
+
+func (r *localRaft) lastIndex() uint64 {
+	return r.raft.LastIndex()
+}
+
+func (r *localRaft) sync(index uint64, timeout time.Duration) error {
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		// Wait for next tick or timeout.
+		select {
+		case <-ticker.C:
+		case <-timer.C:
+			return errors.New("timeout")
+		}
+
+		// Compare index against current metadata.
+		r.store.mu.Lock()
+		ok := (r.store.data.Index >= index)
+		r.store.mu.Unlock()
+
+		// Exit if we are at least at the given index.
+		if ok {
+			return nil
+		}
+	}
+}
+
+func (r *localRaft) snapshot() error {
+	future := r.raft.Snapshot()
+	return future.Error()
+}
+
+// addPeer adds addr to the list of peers in the cluster.
+func (r *localRaft) addPeer(addr string) error {
+	peers, err := r.peerStore.Peers()
+	if err != nil {
+		return err
+	}
+
+	if len(peers) >= 3 {
```

Comment:
Some people will want to run 5-node Raft clusters. How will this code change to support that?